Python/gc.c

```diff
@@ -1558,24 +1547,26 @@ mark_at_start(PyThreadState *tstate)
 static intptr_t
 assess_work_to_do(GCState *gcstate)
 {
-    /* The amount of work we want to do depends on three things.
+    /* The amount of work we want to do depends on two things.
      * 1. The number of new objects created
-     * 2. The growth in heap size since the last collection
-     * 3. The heap size (up to the number of new objects, to avoid quadratic effects)
+     * 2. The heap size (up to twice the number of new objects, to avoid quadratic effects)
+     * 3. The amount of garbage.
      *
-     * For a steady state heap, the amount of work to do is three times the number
-     * of new objects added to the heap. This ensures that we stay ahead in the
-     * worst case of all new objects being garbage.
+     * We cannot know how much of the heap is garbage, but we know that no reachable object
+     * is garbage. We make a (fairly pessismistic) assumption that half the heap not
+     * reachable from the roots is garbage, and count collections of increments as half as efficient
+     * as processing the heap as the marking phase.
      *
-     * This could be improved by tracking survival rates, but it is still a
-     * large improvement on the non-marking approach.
+     * For a large, steady state heap, the amount of work to do is at least three times the
+     * number of new objects added to the heap. This ensures that we stay ahead in the
+     * worst case of all new objects being garbage.
      */
     intptr_t scale_factor = gcstate->old[0].threshold;
     if (scale_factor < 2) {
         scale_factor = 2;
     }
     intptr_t new_objects = gcstate->young.count;
-    intptr_t max_heap_fraction = new_objects*3/2;
+    intptr_t max_heap_fraction = new_objects*5;
```

Comment:
Why is this called `fraction`?

Reply:
Not for any good reason. I'll rename it.